jax/image/scale.py

```diff
@@ -0,0 +1,193 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import enum
+import math
+from typing import Callable, Sequence, Tuple, Union
+
+from jax import lax
+from jax import numpy as jnp
+import numpy as np
+
+
+def _lanczos_kernel(radius: float):
+  def kernel(x: np.ndarray) -> np.ndarray:
+    x = np.abs(x)
+    y = radius * np.sin(np.pi * x) * np.sin(np.pi * x / radius)
+    with np.errstate(divide='ignore', invalid='ignore'):
+      out = y / (np.pi ** 2 * x ** 2)
+    out = np.where(x <= 1e-3, 1., out)
+    return np.where(x > radius, 0., out)
+
+  return radius, kernel
+
+
+def _triangle_kernel():
+  return 1., lambda x: np.maximum(0, 1 - np.abs(x))
+
+
+def _keys_cubic_kernel():
+  # http://ieeexplore.ieee.org/document/1163711/
+  # R. G. Keys. Cubic convolution interpolation for digital image processing.
+  # IEEE Transactions on Acoustics, Speech, and Signal Processing,
+  # 29(6):1153–1160, 1981.
+  def kernel(x: np.ndarray) -> np.ndarray:
+    x = np.abs(x)
+    out = ((1.5 * x - 2.5) * x) * x + 1.
+    out = np.where(x >= 1., ((-0.5* x + 2.5) * x - 4.) * x + 2., out)
+    return np.where(x >= 2., 0., out)
+  return 2., kernel
+
+
+def _compute_spans(input_size: int, output_size: int,
+                   scale: float, translate: float,
+                   kernel: Tuple[float, Callable[[np.ndarray], np.ndarray]],
+                   antialias: bool) -> Tuple[np.ndarray, np.ndarray]:
+  """
+  Computes the locations and weights of the spans of a 1D input image.
+  Returns:
+    A `(starts, weights)` tuple of ndarrays, where `starts` has shape
+    `[output_size]` and `weights` has shape `[output_size, span_size]`.
+  """
+  radius, kernel_fn = kernel
+  inv_scale = 1. / scale
+  kernel_scale = max(inv_scale, 1.) if antialias else 1.
+  span_size = min(2 * int(math.ceil(radius * kernel_scale)) + 1, input_size)
+
+  sample_f = (np.arange(output_size) + 0.5) * inv_scale * (1. - translate)
+  span_start = np.ceil(sample_f - radius * kernel_scale - 0.5)
+  span_start = np.clip(span_start, 0, input_size - span_size)
+  kernel_pos = (span_start - sample_f)[:, None] + np.arange(span_size) + 0.5
+  weight = kernel_fn(kernel_pos / kernel_scale)
+  total_weight_sum = np.sum(weight, axis=1, keepdims=True)
+  weights = np.where(
+      np.abs(total_weight_sum) > 1000. * np.finfo(np.float32).min,
+      weight / total_weight_sum, 0)
+  return span_start.astype(np.int32), weights
+
+
+def _scale_and_translate(x, output_shape, scale, translate, kernel,
```

Comment:
I think we should expose this function too, it's useful sometimes - for example you can use it to pull a region out of an image, or to mimic the "align corners" mode of TF's old resize which is needed in certain Unets I think.

Reply:
Happy to do so, but let's do that in another PR. (I'm mostly concerned about needing to write tests, etc. and I'd like to get this in to unblock users.)